### PR TITLE
NMS-19754: DCB: Add inline SSH configuration capture

### PR DESCRIPTION
**Pull request by dino2gnt**
Adds a `capture: ` DCB DSL statement that can be used to scrape the config out of the SSH session output without pushing it via TFTP.

Tested locally using a dcb script of:
```
await: ]$
send: cat /etc/hosts
await: ]$
send: cat /etc/snmp/snmpd.conf
capture: cat /etc/snmp/snmpd.conf
await: ]$
```
Only the output of `cat /etc/snmp/snmpd.conf` is captured as the configuration in this case.  

Bare `capture: ` with no args will capture everything from that point to the end of the output.  I think both cases are useful.

### External References

* Jira (Issue Tracker): https://opennms.atlassian.net/browse/NMS-19754

Assisted-By: Claude Sonnet 4.6 <noreply@anthropic.com>

### CODE REVIEW

**Comment by marshallmassengill** [comment]
Add:
```
  "attribution": {
    "commit": ""
  },
```
to your ~./claude/settings.json file so you don't pick up the "co-authored" by claude on the commits.  Just put a comment in the PR: "Assisted-by: Claude Code: Sonnet 4.6".  Keeps the commit history clean for authors.

File: features/device-config/retrieval/src/main/java/org/opennms/features/deviceconfig/retrieval/impl/RetrieverImpl.java [hunk 1]
```diff
@@ -104,6 +105,37 @@ public CompletionStage<Either<Failure, Success>> retrieveConfig(
...
+
+                if (scriptResult.isFailed()) {
+                    return Either.<Failure, Success>left(new Failure(
+                            scriptingFailureMsg(target, scriptResult.message),
+                            scriptResult.stdout, scriptResult.stderr, scriptResult.scriptOutput));
+                }
+
+                if (scriptResult.capturedConfig.isEmpty()) {
+                    return Either.<Failure, Success>left(new Failure(
+                            "no config captured - script must include a 'capture:' statement - target: " + target,
```
Comment (marshallmassengill) [inline features/device-config/retrieval/src/main/java/org/opennms/features/deviceconfig/retrieval/impl/RetrieverImpl.java]: The wording here seems off to me.  It's basically an empty response, right?

Maybe something like: 
"capture: statement parsed but produced no output - target: " + target
 or
"internal error: capture: statement did not yield any captured bytes - target: " + target

File: features/device-config/retrieval/src/main/java/org/opennms/features/deviceconfig/retrieval/impl/RetrieverImpl.java [hunk 2]
```diff
@@ -230,6 +262,12 @@ public void onFileReceived(InetAddress address, String fileName, byte[] content)
         }
     }
 
+    static boolean scriptContainsCapture(String script) {
+        return Stream.of(script.split("\\\\n|\\n"))
+                .map(String::trim)
+                .anyMatch(line -> line.equals("capture:") || line.startsWith("capture:"));
```
Comment (marshallmassengill) [inline features/device-config/retrieval/src/main/java/org/opennms/features/deviceconfig/retrieval/impl/RetrieverImpl.java]: You can drop the "line.equals("capture:")" and just use the startsWith, I think?  Seems like one would get the other.

File: features/device-config/ssh-scripting/src/main/java/org/opennms/features/deviceconfig/sshscripting/impl/Statement.java [hunk 3]
```diff
@@ -89,6 +89,34 @@ public void execute(SshInteraction interaction, String string) throws Exception
...
+            /**
+             * {@code capture:} may optionally take a string argument.
+             * With no argument ({@code capture:}), capture starts immediately at the current stdout position.
+             * With an argument ({@code capture: show running-config}), the argument is first awaited
+             * (consuming the device's echo of the preceding send: command), then capture starts —
+             * so the echoed command line is excluded from the captured bytes.
+             */
+            @Override
+            public Statement parse(String line) {
+                if (line.equals(name() + ":")) {
```
Comment (marshallmassengill) [inline features/device-config/ssh-scripting/src/main/java/org/opennms/features/deviceconfig/sshscripting/impl/Statement.java:102]: I asked this above but I don't think you need both line.equals and line.startsWith.  I think you can remove the line.equals path here.

File: features/device-config/retrieval/src/test/java/org/opennms/features/deviceconfig/retrieval/impl/RetrieverImplTest.java [hunk 4]
```diff
@@ -28,8 +28,10 @@
 import static org.junit.Assert.assertThat;
 import static org.junit.Assert.assertTrue;
 import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.anyBoolean;
```
Comment (marshallmassengill) [inline features/device-config/retrieval/src/test/java/org/opennms/features/deviceconfig/retrieval/impl/RetrieverImplTest.java]: This doesn't seem to ever be called.

**Review by marshallmassengill — CHANGES_REQUESTED** [review]

**Comment by marshallmassengill** [comment]
One more thing that showed up when I asked CC about test coverage:

- No test exercises capture: <string> (i.e. awaitAndCapture) in either StatementTest (parse-only there) or RetrieverImplTest — only bare capture: is in the unit tests.
- The unit tests mock the SshScriptingService and pre-bake capturedConfig bytes, so the byte-offset bookkeeping in SshInteractionImpl (captureStart/captureEnd/trim) is not exercised by any automated test. An end-to-end test in SshScriptingServiceImplIT against the existing in-process Apache MINA SSHD echo server would catch the trim/concurrency edges and the prompt-exclusion behavior the docs mention.
- Add at least one IT in SshScriptingServiceImplIT covering both capture: forms end-to-end against the embedded SSHD.

File: features/device-config/ssh-scripting/src/main/java/org/opennms/features/deviceconfig/sshscripting/impl/Statement.java (hunk 3 above)
Comment (dino2gnt) [reply features/device-config/ssh-scripting/src/main/java/org/opennms/features/deviceconfig/sshscripting/impl/Statement.java:102]: We can do as you describe in the instance you pointed put above, but you cannot here.  `capture:` and `capture: <string>` actually behave differently.

**Review by marshallmassengill — DISMISSED** [review]
Changes look good.  There is still the attribution to claude in the first commit.  It may be worth fixing that with a force push but that may also mean you need squash the commits.  I'll let @cgorantla comment on that.

**Comment by dino2gnt** [comment]
> Changes look good. There is still the attribution to claude in the first commit. It may be worth fixing that with a force push but that may also mean you need squash the commits. I'll let @cgorantla comment on that.

Ah.  I never had any intention of fixing that.  Why do we care?

**Comment by marshallmassengill** [comment]
clean commiter history

**Comment by dino2gnt** [comment]
> clean commiter history

If Claude wrote it, all or in part, and Claude committed it, I don't see any reason that Claude shouldn't have a `co-authored-by` tag. It's not dirtying up the commit history, it's keeping it _honest_.

File: features/device-config/retrieval/src/main/java/org/opennms/features/deviceconfig/retrieval/impl/RetrieverImpl.java [hunk 5]
```diff
@@ -104,6 +105,37 @@ public CompletionStage<Either<Failure, Success>> retrieveConfig(
...
+                    return Either.<Failure, Success>left(new Failure(msg));
+                }
+
+                if (scriptResult.isFailed()) {
+                    return Either.<Failure, Success>left(new Failure(
+                            scriptingFailureMsg(target, scriptResult.message),
+                            scriptResult.stdout, scriptResult.stderr, scriptResult.scriptOutput));
+                }
+
+                if (scriptResult.capturedConfig.isEmpty()) {
```
Comment (cgorantla) [inline features/device-config/retrieval/src/main/java/org/opennms/features/deviceconfig/retrieval/impl/RetrieverImpl.java]: May be there are other conditions like whitespace, single character returns like `\n` or \r`.  These cases can be real failures.

**Review by cgorantla — COMMENTED** [review]
One minor thing to fix

It would be good to test this with some real devices and address any corner cases.

File: features/device-config/retrieval/src/main/java/org/opennms/features/deviceconfig/retrieval/impl/RetrieverImpl.java (hunk 5 above)
Comment (dino2gnt) [reply features/device-config/retrieval/src/main/java/org/opennms/features/deviceconfig/retrieval/impl/RetrieverImpl.java]: This is a very good point.

I added a check for printable characters in the captured output, so if all we get back is non-printable chars or whitespace, it should fail.

**Review by cgorantla — APPROVED** [review]
LGTM!
